odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,497 @@
+import odl
+import numpy as np
+standard_library.install_aliases()
+
+
+__all__ = ('linear_deform', 'deform_grad', 'LinDeforFixedTempOp',
+           'LinDeforFixedTempDeriv', 'LinDeforFixedTempDerivAdj',
+           'LinDeforFixedDispOp', 'LinDeforFixedDispAdj')
+
+
+def linear_deform(template, displacement):
+    """Linearized deformation of a template with a displacement field.
+
+    The template ``I`` is deformed by a displacement field ``v`` as:
+
+        (I, v) --> I(Id + v)
+
+    Parameters
+    ----------
+    template : `DiscreteLpVector`
+        Template to be deformed by a displacement field.
+    displacement : `ProductSpaceVector`
+        The vector field (displacement field) used in the linearized
+        deformation.
+
+    Returns
+    -------
+    deformed_template : `numpy.ndarray`
+        Deformed template as an numpy array.
+
+    Examples
+    --------
+    Deform a 1D function ``template`` that has value ``1`` at 3:rd sample
+    point using a displacement field ``displacement_field`` that has
+    value ``-0.2`` in the 4:th sample point. The outcome should be an
+    array with values ``1`` at 3:rd and 4:th positions.
+
+    >>> import odl
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+    >>> linear_deform(template, displacement_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+    """
+    assert template in displacement[0].space
+
+    image_pts = template.space.grid.points()
+    image_pts += np.asarray(displacement).T
+
+    return template.interpolation(image_pts.T, bounds_check=False)
+
+
+def deform_grad(grad_f, displacement):
+    """Compute the deformation of the template gradient.
+
+    The template gradient ``grad(I)`` after deformation with displacement
+    field ``v``: grad(I)(Id + v)
+
+    Parameters
+    ----------
+    grad_f: `ProductSpaceVector`
+        Gradient of the template, i.e. a vector field on the
+        image domain
+    displacement : `ProductSpaceVector`
+        The vector field (displacement field) used in the linearized
+        deformation.
+    Examples
+    --------
+    >>> import odl
+
+    Define ``gradOp`` as the gradient operator, let it act on ``template``.
+
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> gradOp = odl.Gradient(template.space, method='forward')
+    >>> template_grad = gradOp(template)
+    >>> template_grad
+    ProductSpace(uniform_discr(0.0, 1.0, 5), 1).element([
+        [0.0, 5.0, -5.0, 0.0, 0.0]
+    ])
+
+    Define a displacement field ``displacement_field``.
+
+    >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+
+    Calculate linearized deformation of the ``template_grad`` (gradient of
```

Comment:
This type of comment is not what we strive for. In general, code should be self documenting in this sense. By reading the line bellow, any information of this line is quite obviously available to me.

What you should write instead, is *why* this is done, and also *why* we expect this result.